src/prommis/nanofiltration/diafiltration.py

```diff
@@ -0,0 +1,774 @@
+# University of California, through Lawrence Berkeley National Laboratory, et al. All rights reserved.
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and license information.
+#####################################################################################################
+"""
+Diafiltration cascade flowsheet for separating lithium and cobalt. Serves as an example of a
+custom cost model.
+"""
+
+from pyomo.environ import (
+    ConcreteModel,
+    Expression,
+    Objective,
+    Param,
+    Set,
+    Suffix,
+    TransformationFactory,
+    Var,
+    log,
+    units,
+    value,
+)
+from pyomo.network import Arc
+
+import idaes.logger as idaeslog
+from idaes.core import (
+    FlowsheetBlock,
+    MaterialBalanceType,
+    UnitModelBlock,
+    UnitModelCostingBlock,
+)
+from idaes.core.solvers import get_solver
+from idaes.core.util.initialization import propagate_state
+from idaes.core.util.model_diagnostics import DiagnosticsToolbox
+from idaes.core.util.model_statistics import degrees_of_freedom
+from idaes.models.unit_models import (
+    Mixer,
+    MixerInitializer,
+    MixingType,
+    MomentumMixingType,
+    MSContactor,
+    MSContactorInitializer,
+)
+
+from prommis.nanofiltration.costing.diafiltration_cost_model import (
+    DiafiltrationCosting,
+    DiafiltrationCostingData,
+)
+from prommis.nanofiltration.diafiltration_properties import LiCoParameters
+
+_log = idaeslog.getLogger(__name__)
+
+
+def main():
+    """
+    Builds and solves the diafiltration flowsheet with cost
+
+    Returns:
+        m: Pyomo model
+    """
+    m = build_model()
+    if degrees_of_freedom(m) != 0:
+        raise ValueError(
+            "Degrees of freedom were not equal to zero after building model"
+        )
+
+    add_costing(m)
+    if degrees_of_freedom(m) != 0:
+        raise ValueError(
+            "Degrees of freedom were not equal to zero after building cost block"
+        )
+
+    initialize_model(m)
+    _log.info("Initialization Okay")
+
+    dt = DiagnosticsToolbox(m)
+    dt.assert_no_structural_warnings(ignore_evaluation_errors=True)
+
+    solve_model(m)
+    _log.info("Solved Square Problem")
+
+    dt.assert_no_numerical_warnings()
+
+    unfix_opt_variables(m)
+    add_product_constraints(m, Li_recovery_bound=0.95, Co_recovery_bound=0.4)
+    add_objective(m)
+
+    # Create a scaled version of the model to solve
+    set_scaling(m)
+    scaling = TransformationFactory("core.scale_model")
+    scaled_model = scaling.create_using(m, rename=False)
+    solve_model(scaled_model)
+    # Propagate results back to unscaled model
+    scaling.propagate_solution(scaled_model, m)
+
+    dt.assert_no_numerical_warnings()
+    print_information(m)
+
+    return m
+
+
+def build_model():
+    """
+    Builds the diafiltration flowsheet
+
+    Returns:
+        m: Pyomo model
+    """
+    m = ConcreteModel()
```

Comment:
Remove this line so that `m` refers to an existing model object. The method would take `m` as an input, like `def build_model(m)`.